### ISSUE TITLE
I have a present for you

### ISSUE THREAD

**Issue by kdschlosser** [WoLpH/python-progressbar#223]
I give to you the gift of detecting ANSI support

```python
import sys
import os
import time

sample_ansi = '\x1b[31mRED ' + '\x1b[33mYELLOW ' + '\x1b[32mGREEN ' + '\x1b[35mPINK ' + '\x1b[0m' + '\n'

handle = sys.stdout

if (
    # This works for newer versions of pycharm only. older versions there is no way to check.
    ('PYCHARM_HOSTED' in os.environ and os.environ['PYCHARM_HOSTED'] == '1') or
    (
        # check if we are writing to a terminal or not. typically a file object is going to return False
        # if the instance has been overridden and isatty has not been defined we have no way of knowing
        # so we will not use ansi.
        (hasattr(handle, "isatty") and handle.isatty()) and
        (
            # ansi terminals will typically define one of the 2 environment variables.
            ('TERM' in os.environ and os.environ['TERM'] == 'ANSI') or
            'ANSICON' in os.environ
        )
    )
):
    handle.write("ANSI output enabled.\n")
    handle.write(sample_ansi)

elif sys.platform.startswith('win'):
    handle.write("Windows console, no ANSI support.\n")

else:
    handle.write('ANSI output disabled.\n')

handle.write("\n\n")
handle.flush()
```

This will detect if a 3rd party terminal that supports ANSI is being used on Windows.
It also detects if PyCharm is being used. (latest version of PyCharm has been tested and does work)

**Comment by WoLpH** [comment]
Thank you, that's very useful to know. Especially detecting Pycharm could help in adding some preliminary warnings of the Pycharm console handling.

This would be a good way to detect ANSI support on Windows perhaps. On Linux/OS X it probably needs to be expanded a little but it would be nice to add something like this. Automatic color support would be nice :)
For now I'm thinking adding `xterm` and `xterm-256color` 
I should note that the ANSI detection is a bit too limited for now. My terminal definitely supports ANSI colors but the test is failing. I think it would need at least support for `xterm.*`, `screen.*` and `rxvt` being the `TERM` variable.

I'm a bit scared of always enabling those as these types of changes have broken a lot in the past. But for the very specific cases it could be very useful indeed :)

**Comment by kdschlosser** [comment]
Here is what I have set up.

This should be a really good layout for testing.

```python
# -*- coding: utf-8 -*-

import os
import sys


def is_ansi_terminal(handle):
    if (
        # This works for newer versions of pycharm only. older versions there is no way to check.
        (
            'PYCHARM_HOSTED' in os.environ and
            os.environ['PYCHARM_HOSTED'] == '1' and
            handle in (sys.__stdout__, sys.__stderr__)
        ) or (
            # check if we are writing to a terminal or not. typically a file object is going to return False
            # if the instance has been overridden and isatty has not been defined we have no way of knowing
            # so we will not use ansi.
            (hasattr(handle, "isatty") and handle.isatty()) and (
                # ansi terminals will typically define one of the 2 environment variables.
                ('TERM' in os.environ and os.environ['TERM'] == 'ANSI') or
                'ANSICON' in os.environ
            )
        )
    ):
        return True

    return False


def is_windows_terminal(handle):
    return (
        not is_ansi_terminal(handle) and
        (
            sys.platform.startswith('win') and
            hasattr(handle, "isatty") and
            handle.isatty()
        )
    )
```

use

```python
import sys

handle = sys.stdout

if is_ansi_terminal(handle):
    # do ANSI code here
    pass
elif is_windows_terminal(handle):
    # do windows CMD shell code here
    pass
else:
    # This is for all other file like objects.
```

I had to add some additional checks for PyCharm. Because a PyCharm terminal isatty() is going to return False. and we do not want it to be treated as a file like object if that is the case.

sys.__stdout__ and sys.__stderr__ are the original stdout and stderr that are made when the interpreter starts. so we can check and see if the file like object being passed is actually one of those 2. and if it is and the environment has the PYCHARM_HOSTED set we then know it is going to be an  ANSI terminal even tho isatty() returns False.

I have spent a little bit of time working on modifying you library. I think you are going to like the changes made. It makes the library 100% thread safe, handles the terminal types correctly, colors can be set for ANSI and Windows console, It updates the bar based on row number. so a user can have more then one bar running at the same time from multiple threads.

I am running through the initial testing of it to get as many of the kinks out of it as I can.

Also I put in a PR for your python-utils for a remap feature. this would be a handy tool for remaping a supplied min & max to the starting & ending column for the bar so you will know exactly what needs to be "filled". It has a great use case for the progress bhar but it also has a plethora of other uses as well. 


I do not know how Jupyter handles isatty() if it returns Ture or False you may want to check that just in case it returns False. you would want to do the same kind of a test that I do for PyCharm if it does.

**Comment by kdschlosser** [comment]
And the thread safe bit also apply when creating multiple bars and passing the same file object to all of the bars. I created a wrapper around the file object and only a single wrapper can exist for each file object. so each bar that gets created will share the same wrapper to handle the IO.

I also wrote in a non blocking stdin reader to be able to collect the response for the ANSI getpos. this is cross platform too, Windows, OSX, Linux

**Comment by kdschlosser** [comment]
Here is a better version if the is_ansi_terminal

```python
# -*- coding: utf-8 -*-

import os
import sys
import re


ANSI_TERMS = (
    '([xe]|bv)term',
    '(sco)?ansi',
    'cygwin',
    'konsole',
    'linux',
    'rxvt',
    'screen',
    'tmux',
    'vt(10[02]|220|320)',
)
ANSI_TERM_RE = re.compile('^({})'.format('|'.join(ANSI_TERMS)), re.IGNORECASE)


def is_ansi_terminal(handle):
    is_ansi = False
    
    if handle in (sys.__stdout__, sys.__stderr__):
        # This works for newer versions of pycharm only. older versions there 
        # is no way to check.
        if (
            'PYCHARM_HOSTED' in os.environ and 
            os.environ['PYCHARM_HOSTED'] == '1'
        ):
            is_ansi = True
            
        elif 'JPY_PARENT_PID' in os.environ:
            is_ansi = True
            
    # check if we are writing to a terminal or not. typically a file object is 
    # going to return False if the instance has been overridden and isatty has 
    # not been defined we have no way of knowing so we will not use ansi.
    # ansi terminals will typically define one of the 2 environment variables.
            
    if (
        hasattr(handle, "isatty") and handle.isatty() and (
            ANSI_TERM_RE.match(os.environ.get('TERM', '')) or 
            'ANSICON' in os.environ
        )
    ):
        is_ansi = True

    return is_ansi


def is_windows_terminal(handle):
    return (
        not is_ansi_terminal(handle) and
        (
            sys.platform.startswith('win') and
            hasattr(handle, "isatty") and
            handle.isatty()
        )
    )

```

**Comment by kdschlosser** [comment]
I just opened an issue on PyCharms issue tracker. The get position ANSI command does not work. I also asked for a complete list of all supported ANSI commands from them.

**Comment by WoLpH** [comment]
Thank you for the extensive information and all of the suggestions! They're really useful :)

I'm wondering if the `sys.__stout__` check is needed actually, that test would fail in some scenarios since there are libraries that actually override it (even though they shouldn't) so I'm not sure it would be useful to put it in this project.

As for Jupyter, when I last tested they did not return true for `isatty` but the `JPY_PARENT_ID` seemed a useful workaround. But it's probably far from perfect and might also function in the wrong cases. Ideally I want to add a Jupyter Widget in the future.

I think I've added all of your changes now and I'll create a new release for the python utils as well.

Once again thank you for all the help!

**Comment by kdschlosser** [comment]
Hey no worries m8. I am glad to help.

The reason why I bring up the Jupyter check is it then has the same very fatal flaw as PyCharm
Here is a very real scenario.

Running a PyCharm or Jupyter terminal.

```python

import progressbar

output_stream = open('some_file.txt', 'w')
pb = progressbar.ProgressBar(fd=output_stream)
```

Now there is a problem. even tho the object passed is not a terminal it is going to be treated like it is one and thus ANSI escape characters are going to get written.

Now if memory serves I believe that
stdin.fileno() = 0
stdout.fileno() = 1
stderr.fileno() = 3

I think this is a constant. So if some other number gets returned or fileno() is not implemented then you know beyond a reasonable doubt that those are in fact the original IO handlers.

I know that sys.stdout and sys.stderr get overwritten all the time. sys.\_\_stdout\_\_ and sys._\_\stderr\_\_ on the other hand are left alone for the most part. there is a rare occasion where these are overridden but not that often. 

in your DefaultFdMixin class constructor you have the following lines of code
```python
        if fd is sys.stdout:
            fd = utils.streams.original_stdout

        elif fd is sys.stderr:
            fd = utils.streams.original_stderr
```

This seems odd to me. because progressbar can be imported and stdout and stderr can be overridden after that happens. but by the design here the progress bar is always going to use the "originals" and on the other hand there is a default parameter set for the fd parameter of sys.stderr. This default is going to capture the "real one" and if a user overrides them after progressbar gets imported that is also going to cause some strange behavior.

The above is more questions then anything. I am trying to understand how the mechanics work.


You can verify the authenticity of sys.stdout and sys.stderr in Windows by using the following code.
```python
import sys

import msvcrt

import ctypes
from ctypes.wintypes import (
    BOOL,
    HANDLE,
    DWORD,
    _COORD,
    WORD,
    SMALL_RECT,
)

ENABLE_VIRTUAL_TERMINAL_PROCESSING = 0x0004
STD_OUTPUT_HANDLE = -11
STD_ERROR_HANDLE = -12

COORD = _COORD
POINTER = ctypes.POINTER

kernel32 = ctypes.windll.Kernel32

GetStdHandle = kernel32.GetStdHandle
GetStdHandle.restype = HANDLE

print(msvcrt.get_osfhandle(sys.stdout.fileno()))
print(GetStdHandle(STD_OUTPUT_HANDLE))
```

There is probably something similiar for posix as well.

**Comment by WoLpH** [comment]
> Now there is a problem. even tho the object passed is not a terminal it is going to be treated like it is one and thus ANSI escape characters are going to get written.
> 
> Now if memory serves I believe that
> stdin.fileno() = 0
> stdout.fileno() = 1
> stderr.fileno() = 3
> 
> I think this is a constant. So if some other number gets returned or fileno() is not implemented then you know beyond a reasonable doubt that those are in fact the original IO handlers.
> 
> I know that sys.stdout and sys.stderr get overwritten all the time. sys.__stdout__ and sys.__\stderr__ on the other hand are left alone for the most part. there is a rare occasion where these are overridden but not that often.

You're right, that's definitely an issue. The big problem with changing that behaviour now is that it might break existing code for people. Especially when it comes to Jupyter Notebooks some are used for a long time and if I change the behaviour I could break things for users...


> in your DefaultFdMixin class constructor you have the following lines of code
> 
> ```python
>         if fd is sys.stdout:
>             fd = utils.streams.original_stdout
> 
>         elif fd is sys.stderr:
>             fd = utils.streams.original_stderr
> ```
> 
> This seems odd to me. because progressbar can be imported and stdout and stderr can be overridden after that happens. but by the design here the progress bar is always going to use the "originals" and on the other hand there is a default parameter set for the fd parameter of sys.stderr. This default is going to capture the "real one" and if a user overrides them after progressbar gets imported that is also going to cause some strange behavior.

That's one of the hacks needed to make sure the output redirection works properly. In case you want to mix using a progressbar while still being able to write to stdout and stderr without getting interleaved lines, that is part of the hack that solves the issue.

One library of where this is an issue is the `logging` library that stores `stderr` on import: https://github.com/python/cpython/blob/master/Lib/logging/__init__.py#L1038-L1056
https://github.com/python/cpython/blob/e0acec15412203359d59db33e447698ee51e07fc/Lib/logging/__init__.py#L1192-L1209

> There is probably something similiar for posix as well.

The only thing I can think of is `isatty()` with `fileno()` combined.

One of the biggest issues for this library is backwards compatible as it's currently still possible to use as a drop-in replacement for the original version of `progressbar` that was released about 15 years ago. Since it's also included in major Linux/Unix distro's I need to make sure the behaviour stays as constant as possible :)

**Comment by kdschlosser** [comment]
I didn't know this library has been around for that long. and gets used in the manner that it does So i do understand the need to keep the API and functionality in tact.

But moving forward with the addition of PyCharm compatibility I would suggest making sure the same issue that exists with Jupyter does not exist with PyCharm. so if you do a check of isatty() returning False 'PYCHARM_HOSTED' is in the environment and the fileno() is either 1 or 2 and use the mechanism above to double check the handle the system has for the fd and if it matches the systems handle for stdout or stderr then you know beyond any kind of a doubt that it is in fact a terminal. 

The method I showed above for windows and checking the handle is something that should be done to verify stdout and stderr as the terminal, that method works for ansi terminal, cmd shell and PyCharm terminal. I have tested it.


> That's one of the hacks needed to make sure the output redirection works properly. In case you want to mix using a progressbar while still being able to write to stdout and stderr without getting interleaved lines, that is part of the hack that solves the issue.

Now I understand what the stream capture is for. and now it also makes sense why you said adding in thread safety and cursor movements for doing the drawing would be a lot of work.

Basically the existing progress bar classes would have to remain the same as they are currently, a new set of classes would need to be made in order to facilitate those changes. I was running into a problem that I was not sully understanding where I am not able to display more then a single bar at a time using the current classes. This is happening because of that mechanism.

**Comment by WoLpH** [comment]
> I didn't know this library has been around for that long. and gets used in the manner that it does So i do understand the need to keep the API and functionality in tact.

This fork only exists since 2012, but the original abandoned project existed for quite a few years before that.

It's hard to research how many people are using this package versus the original package but it's in the repository systems for several distros. And since it can be used as a drop-in replacement people shouldn't really notice when they've upgraded ;)

> But moving forward with the addition of PyCharm compatibility I would suggest making sure the same issue that exists with Jupyter does not exist with PyCharm. so if you do a check of isatty() returning False 'PYCHARM_HOSTED' is in the environment and the fileno() is either 1 or 2 and use the mechanism above to double check the handle the system has for the fd and if it matches the systems handle for stdout or stderr then you know beyond any kind of a doubt that it is in fact a terminal.

As far as I know the `isatty()` check actually already worked with Pycharm, although it may depend on the settings.

Specifically the "Enable terminal in output console" setting: 
![image](https://user-images.githubusercontent.com/270571/75829311-1e2a4200-5dae-11ea-8fb8-9b4e057a0bdf.png)

I'm going to do some testing to make sure it works well.

> The method I showed above for windows and checking the handle is something that should be done to verify stdout and stderr as the terminal, that method works for ansi terminal, cmd shell and PyCharm terminal. I have tested it.

Excellent! :)
I'll have to make sure it still works with nested/multiple progressbar situations and such but that's good to know.

**Comment by kdschlosser** [comment]
It would have been nice if Python had set the "real" stderr, stdout and stdin into a read only variable of some kind. This way it could always be found. I m sure that is what \_\_stdout\_\ and \_\_stderr\_\_ was intended to be. Maybve it is best to use it in that manner, and if someone has overridden those and ends up with a problem when using the library they can then be instructed that overriding \_\_stdout\_\_ and \_\_stderr\_\_ is not supposed to be done.

I am going to check if there is an equivalent to GetStdHandle for posix systems. This would provide a sure fire method of detecting stdout or stderr.

Also I wanted to pass on a bit of code so you can test it for me on a posix system. This is a non blocking stdin reader. It should be cross platform between Linux, OSX and Windows. You will just need to run it. It should spit out the cursor position. This will not ork from inside of PyCharm. PyCharm does not have support for the full set of ANSI codes. I think it only has support for colors. I have asked them for a list of supported ANSI codes.

```python
# -*- coding: utf-8 -*-

from __future__ import print_function
import sys

try:
    # Windows
    import msvcrt

    def _getch():
        return msvcrt.getch().decode('utf-8')

    def _kbhit():
        return msvcrt.kbhit()

    def end_read():
        pass

    def start_read():
        pass

except ImportError:
    # Posix (Linux, OS X)
    import termios
    import atexit
    from select import select

    fd = sys.stdin.fileno()

    # Save the terminal settings
    old_term = termios.tcgetattr(fd)

    def end_read():
        termios.tcsetattr(fd, termios.TCSAFLUSH, old_term)
        atexit.unregister(end_read)

    def _getch():
        return sys.stdin.read(1)

    def _kbhit():
        dr, dw, de = select([sys.stdin], [], [], 0)
        return dr != []


    def start_read():
        new_term = termios.tcgetattr(fd)

        # New terminal setting unbuffered
        new_term[3] = (new_term[3] & ~termios.ICANON & ~termios.ECHO)
        termios.tcsetattr(fd, termios.TCSAFLUSH, new_term)

        # Support normal-terminal reset at exit
        atexit.register(end_read)


def read():
    data = ''

    while _kbhit():
        data += _getch()

    return data


# Test
if __name__ == "__main__":

    import time
    print('Testing for ANSI escape return value for cursor position.')
    print('sending ANSI code ' + repr('\x1B[6n'))
    print('Hit any key, or ESC to exit')

    start_read()
    sys.stdout.write('\x1B[6n')
    sys.stdout.flush()

    while True:
        in_data = read()
        if in_data:
            if len(in_data) == 1 and ord(in_data) == 27:  # ESC
                break

            print(repr(in_data))
        else:
            time.sleep(0.1)

    end_read()
```

**Comment by kdschlosser** [comment]
OK so in posix systems the terminal handles are 
0 = stdin
1 = stdout
2 = stderr

The nice thing about this is if someone uses a command like this
```python test_script.py > test_output.log```

the handles gotten by using stdout.fileno() and stderr.fileno() shouldn't be 1 and 2. This might be something you want to check out to see if that is the case. That would be nice to have so you would know if the output is being ported to a file and your program would be able to adjust accordingly.

**Comment by kdschlosser** [comment]
OK This is still a work in progress. I do not believe I changed any of the public API at all.

https://github.com/kdschlosser/python-progressbar/tree/threads_test

run `test.py` from an ANSI terminal to give it a go. 

DO NOT run this from inside of PyCharm. PyCharm has no support for moving the cursor position. I have not added code to disallow use inside of PyCharm.

I have not dove into the wrappers or anything like that. You might want to run some tests against the thing and tell me what is and is not working properly. The test I have written uses the ProgressBar class. I do not know if all of the other classes subclass the DefaultFdMixin class or not. If they do then everything should work.

I added color constants to the program as well. These are not fully implemented yet. There would need to be a modification made to the public API in order to set the colors/text attributes. The color constants can be joined together using bitwise or (|). I added the ability to set the color/attributes to be global for an fd or when something is going to get written. passing the color/attributes when writing overrides the global settings. 

This is about the cleanest way I could think of adding the multi bar support as well as making the library thread safe.

I am sure there are things you may want to be set up differently, all of the mechanics are there and you are more then welcome to use what you want and arrange it however you feel it should be done. You should easily get an understanding of what is going on with the changes I have made.

It should work properly for ANSI terminals Windows included and also the Windows cmd shell.

If you have any questions fire away.
I am hopeful that you like what I have done in part or in whole.

**Comment by WoLpH** [comment]
The code looks good and definitely like something that I would love to add.
As of now it doesn't appear to be working from my terminal. Which is definitely ANSI compliant.

In Python 2.7 I get no output until I press the enter button, after which I see the following stacktraces:
```

# python test.py
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.17/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python@2/2.7.17/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "test.py", line 9, in do
    pb = progressbar.ProgressBar(min_value=0, max_value=100).start()
  File "progressbar/bar.py", line 707, in start
    StdRedirectMixin.start(self, max_value=max_value)
  File "progressbar/bar.py", line 178, in start
    DefaultFdMixin.start(self, *args, **kwargs)
  File "progressbar/bar.py", line 118, in start
    self._row = self.fd.y
  File "progressbar/base.py", line 186, in y
    position = self.get_position()
  File "progressbar/ansi_cursor.py", line 85, in get_position
    stdin.end_read()
  File "progressbar/stdin.py", line 35, in end_read
    atexit.unregister(end_read)
AttributeError: 'module' object has no attribute 'unregister'

Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.17/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python@2/2.7.17/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "test.py", line 9, in do
    pb = progressbar.ProgressBar(min_value=0, max_value=100).start()
  File "progressbar/bar.py", line 707, in start
    StdRedirectMixin.start(self, max_value=max_value)
  File "progressbar/bar.py", line 178, in start
    DefaultFdMixin.start(self, *args, **kwargs)
  File "progressbar/bar.py", line 118, in start
    self._row = self.fd.y
  File "progressbar/base.py", line 186, in y
    position = self.get_position()
  File "progressbar/ansi_cursor.py", line 85, in get_position
    stdin.end_read()
  File "progressbar/stdin.py", line 35, in end_read
    atexit.unregister(end_read)
AttributeError: 'module' object has no attribute 'unregister'
```

On Python 3.8 I don't get any output either until I press a button. After pressing enter I get:
```
# python test.py
Exception in thread Thread-1:
Traceback (most recent call last):
  File "python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "test.py", line 9, in do
    pb = progressbar.ProgressBar(min_value=0, max_value=100).start()
  File "progressbar/bar.py", line 707, in start
    StdRedirectMixin.start(self, max_value=max_value)
  File "progressbar/bar.py", line 178, in start
    DefaultFdMixin.start(self, *args, **kwargs)
  File "progressbar/bar.py", line 118, in start
    self._row = self.fd.y
  File "progressbar/base.py", line 186, in y
    position = self.get_position()
  File "progressbar/ansi_cursor.py", line 89, in get_position
    y, x = list(int(item) for item in in_data[2:-1].split(';'))
  File "progressbar/ansi_cursor.py", line 89, in <genexpr>
    y, x = list(int(item) for item in in_data[2:-1].split(';'))
ValueError: invalid literal for int() with base 10: ''
Exception in thread Thread-2:
Traceback (most recent call last):
  File "python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "test.py", line 9, in do
    pb = progressbar.ProgressBar(min_value=0, max_value=100).start()
  File "progressbar/bar.py", line 707, in start
    StdRedirectMixin.start(self, max_value=max_value)
  File "progressbar/bar.py", line 178, in start
    DefaultFdMixin.start(self, *args, **kwargs)
  File "progressbar/bar.py", line 118, in start
    self._row = self.fd.y
  File "progressbar/base.py", line 186, in y
    position = self.get_position()
  File "progressbar/ansi_cursor.py", line 89, in get_position
    y, x = list(int(item) for item in in_data[2:-1].split(';'))
  File "progressbar/ansi_cursor.py", line 89, in <genexpr>
    y, x = list(int(item) for item in in_data[2:-1].split(';'))
ValueError: invalid literal for int() with base 10: ''
```

**Comment by kdschlosser** [comment]
This is one of those things I need to come up with some kind of an exit from the loop.

your terminal does not support the full set of ANSI codes. specifically the getpos code to get the cursor position.  I am not sure if the code is a VT100 only code or not. I will have to check into that
```\x1B[6n```

which is supposed to return
```\x1B[{y};{x}R```
x = column number 
y = row number

via stdin. Now stdin is an animal in it's self. Because it will block until some data is received. which we do not want to get stuck waiting for input. The reason you are getting stuck now is because I didn't code in a timeout for the waiting for the reply. This is the problem I have run into with PyCharm.

Are you running on a Linux box? If so check your terminal emulation and set it to VT100. if you can. If not there are a plethora of terminal programs out there I am sure there is one that would work. If you are running from Windows or can run from Windows ConEmu works.

I am going to do some more digging into the getpos code and see if there are other variants of it.

**Comment by WoLpH** [comment]
I'm guessing the output is slightly different for some reason.  This is the response from the `repr()` of `in_data` when pressing <kbd>enter</kbd> a few times:
```
'[2;1R\n'
...
'[6;1R\n'
...
'[28;1R\n'
```

I'm running on OS X with iTerm2 and zsh. Which is pretty much the standard as far as I am aware. I've also tried the Terminal app that's bundled with OS X but the results are the same.
Changing the `TERM` to `vt100` or `ansi` doesn't make a difference. It's `xterm-256color` by default.

When running `xterm` I simply don't get any output at all.

**Comment by kdschlosser** [comment]
give it a try now. I made some modifications to stdin.py hopefully it will now work on posix systems

**Comment by kdschlosser** [comment]
also try just running the stdin.py file. there is a test program  written into it.

**Comment by WoLpH** [comment]
As feared, the `is_terminal` changes caused a regression: https://github.com/WoLpH/python-progressbar/issues/224

Terminal detection remains tricky :)

Also note that I haven't forgotten about your updates, I simply haven't had time :(

**Comment by kdschlosser** [comment]
Oh no worries. This is a testing phase.

OK this should hopefully be the best way to test for a terminal.
This will test for a terminal using the POSIX standard stderr and stdout file numbers.
It also converts the passed objects file number into a windows handle and tests to see if
it is the same handle used by windows for stdout or stderr. This should be a rock solid way of checking. I know that a user can manipulate the fileno method and return whatever number it is they want. I do not think this is something that would get done at all.

I also added in another function. This function specifically tests to see if a terminal is a Windows CMD shell. This is an important check for several reasons. We want to make sure that a shell is running a Windows Terminal if we are going to use Windows API command to control the shell for 1. But for 2 it is used in ANSI detection for a Windows 10 CMD shell. in Windows 10 the CMD shell does support ANSI. This check is important because if we are not able to detect an ANSI shell running on Windows 10 and they are not running the CMD shell then we will want to not use any of the added features of ANSI or the Windows API commands. 

This is everything I know about how to detect a terminal and for Windows figuring out what the shell is.

<details>
  <summary>Example Code</summary>

'''python
# -*- coding: utf-8 -*-

import os
import sys
import re
import platform


ANSI_TERMS = (
    '([xe]|bv)term',
    '(sco)?ansi',
    'cygwin',
    'konsole',
    'linux',
    'rxvt',
    'screen',
    'tmux',
    'vt(10[02]|220|320)',
)
ANSI_TERM_RE = re.compile('^({})'.format('|'.join(ANSI_TERMS)), re.IGNORECASE)


def is_ansi_terminal(handle):
    is_ansi = False

    if handle in (sys.__stdout__, sys.__stderr__):
        # This works for newer versions of pycharm only. older versions there
        # is no way to check.
        if (
            'PYCHARM_HOSTED' in os.environ and
            os.environ['PYCHARM_HOSTED'] == '1'
        ):
            is_ansi = True

        elif 'JPY_PARENT_PID' in os.environ:
            is_ansi = True

    # check if we are writing to a terminal or not. typically a file object is
    # going to return False if the instance has been overridden and isatty has
    # not been defined we have no way of knowing so we will not use ansi.
    # ansi terminals will typically define one of the 2 environment variables.

    if (
        hasattr(handle, "isatty") and handle.isatty() and (
            ANSI_TERM_RE.match(os.environ.get('TERM', '')) or
            'ANSICON' in os.environ
        )
    ):
        is_ansi = True

    if (
        sys.platform.startswith('win') and 
        platform.version().startswith('10') and 
        is_windows_cmd_shell(handle)
    ):
        is_ansi = True

    return is_ansi


def is_terminal(handle):
    if not hasattr(handle, 'fileno'):
        return False

    if sys.platform.startswith('win'):
        # for Windows we need to convert the fileno into a Windows HANDLE, then we check
        # that HANDLE against the standard Windows handles for stdout and stderr to see if 
        # there is a match. If there is a match then it is a terminal  
        import ctypes
        import msvcrt
        from ctypes.wintypes import HANDLE, DWORD, BOOL

        STD_OUTPUT_HANDLE = -11
        STD_ERROR_HANDLE = -12

        kernel32 = ctypes.windll.Kernel32

        GetStdHandle = kernel32.GetStdHandle
        GetStdHandle.restype = HANDLE

        out_handle = GetStdHandle(STD_OUTPUT_HANDLE)
        err_handle = GetStdHandle(STD_ERROR_HANDLE)

        handle_handle = msvcrt.get_osfhandle(handle.fileno())

    else:
        # for POSIX systems the stdout and stderr file numbers are 1 and 2. so we check to see
        # if the passed file object has one of those file numbers if it does then it is a terminal.
        err_handle = 2
        out_handle = 1

        handle_handle = handle.fileno()

    return handle_handle in (err_handle, out_handle)


def is_windows_cmd_shell(handle):
    # This functions is to specifically check to see if a terminal is a windows CMD shell.
    # we do this by querying one of the Windows CMD API functions. if the function returns True
    # then we know that we are running in a Windows CMD shell. The reason for this test
    # is because there are many shell programs that can be used and not all of them support
    # ANSI commands. Windows 10 CMD shell does support ANSI commands. 
    if (
        sys.platform.startswith('win') and
        is_terminal(handle)
    ):
        import ctypes
        import msvcrt
        from ctypes.wintypes import HANDLE, DWORD, BOOL

        kernel32 = ctypes.windll.Kernel32

        class _CONSOLE_CURSOR_INFO(ctypes.Structure):
            _fields_ = [
                ('dwSize', DWORD),
                ('bVisible', BOOL)
            ]

        CONSOLE_CURSOR_INFO = _CONSOLE_CURSOR_INFO

        # BOOL WINAPI GetConsoleCursorInfo(
        #   _In_  HANDLE               hConsoleOutput,
        #   _Out_ PCONSOLE_CURSOR_INFO lpConsoleCursorInfo
        # );
        GetConsoleCursorInfo = kernel32.GetConsoleCursorInfo
        GetConsoleCursorInfo.restype = BOOL

        return GetConsoleCursorInfo(
            HANDLE(msvcrt.get_osfhandle(handle.fileno())), 
            ctypes.byref(CONSOLE_CURSOR_INFO())
        )

    return False
```

</details>
